fuel-poa-coordinator/src/service.rs

```diff
@@ -0,0 +1,326 @@
+        self.import_block_events_tx
+            .send(ImportBlockBroadcast::PendingFuelBlockImported {
+                block: Arc::new(block),
+            })
+            .expect("Failed to import the generated block");
+
+        // Update last block time
+        self.last_block_created = Instant::now();
+
+        // Set timer for the next block
+        match self.trigger {
+            Trigger::Never => {
+                unreachable!("This mode will never produce blocks");
+            }
+            Trigger::Instant => {}
+            Trigger::Interval { block_time } => {
+                // TODO: instead of sleeping for `block_time`, subtract the time we used for processing
+                self.timer.set_timeout(block_time, OnConflict::Min).await;
+            }
+            Trigger::Hybrid {
+                max_block_time,
+                min_block_time,
+                max_tx_idle_time,
+            } => {
+                self.timer
+                    .set_timeout(max_block_time, OnConflict::Min)
+                    .await;
+
+                // TODO: Maybe there could be a way to get the length without having
+                //       to perform full block production computation?
+                let block_tx_count =
+                    self.signal_produce_block().await?.transactions.len();
+                let includable_tx_count = self.txpool_sender.includable().await?.len();
+
+                // If txpool still has more than a full block of transactions available,
+                // produce new block in min_block_time.
+                if includable_tx_count > block_tx_count {
+                    self.timer
+                        .set_timeout(min_block_time, OnConflict::Min)
+                        .await;
+                } else if includable_tx_count > 0 {
+                    // If we still have available txs, reduce the timeout to max idle time
+                    self.timer
+                        .set_timeout(max_tx_idle_time, OnConflict::Min)
+                        .await;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn on_txpool_event(
+        &mut self,
+        txpool_event: &TxStatusBroadcast,
+    ) -> anyhow::Result<()> {
+        match txpool_event.status {
+            TxStatus::Submitted => match self.trigger {
+                Trigger::Instant => {
+                    self.produce_block().await?;
+                    Ok(())
+                }
+                Trigger::Never | Trigger::Interval { .. } => Ok(()),
+                Trigger::Hybrid {
+                    max_tx_idle_time,
+                    min_block_time,
+                    ..
+                } => {
+                    // Create a new block just to see if we can fit them all in
+                    // TODO: Maybe there could be a way to get the length without having
+                    //       to perform full block production computation?
+                    let block_tx_count =
```

Comment:
See comment above - also this logical check on the txpool size vs block capacity should be deduplicated into a different function.

Reply:
Noted, continuing discussion in the item above